_posts/2016-02-29-avoid-runtime-errors-in-react-with-flow.markdown

```diff
@@ -0,0 +1,372 @@
+
+{% highlight javascript %}
+// ...
+import type { Todo } from './TodoListItem'
+
+type Props = {
+  todos: Array<Todo>,
+};
+class TodoList extends React.Component {
+  props: Props;
+
+  @autobind
+  _handleCompleteTodo(id: string): void {
+    // do whatever it was we were doing before
+  }
+
+  render(): React.Element {
+    return (
+      <ul>
+        {this.props.todos.map((todo) => (
+          <li key={todo.id}>
+            <TodoListItem
+              todo={todo}
+              onComplete={this._handleCompleteTodo} // <-- important
+            />
+          </li>
+        ))}
+      </ul>
+    )
+  }
+}
+{% endhighlight %}
+
+Notice that we are no longer creating a function with each iteration of `this.props.todos`. Whether or not we do this with `.bind` or an arrow function is beside the point; the fact is, we are now delegating this responsibility to `<TodoListItem />`. Let's dive in and see what that the modified component would look like (removing some demo code for succinctness):
+
+{% highlight javascript %}
+// ...
+type Props = {
+  todo: Todo,
+  onComplete: Function,
+}
+class TodoListItem extends React.Component {
+  props: Props;
+
+  // Ok, so when a user clicks the button to complete this
+  // todo, we invoke the `onComplete` callback provided by
+  // our parent component, passing it our todo so it knows
+  // which todo was completed.
+  @autobind
+  _handleComplete(): void {
+    this.props.onComplete(this.props.todo)
+  }
+
+  render(): React.Element {
+    return (
+      <div>
+        <button onClick={this._handleComplete}>Complete</button>
+      </div>
+    )
+  }
+}
+{% endhighlight %}
+
+Looks great, except there's a bug, which is... not so great. The `<TodoList />` component's `_handleCompleteTodo` method is still expecting an argument `id` of type `string`, not the entire todo object. What we have right now is still fairly equivalent to what is possible with regular `PropTypes`, meaning `Function` does not say anything about its signature or return type. Luckily, we can go a step further with Flow.
+
+Since we know exactly how our component plans to call `onComplete`, we can explicitly define that signature in the `Props` type definition, which will require any implementers to abide by that signature. Check it out:
+
+{% highlight javascript %}
+type Props = {
+  todo: Todo,
+  onComplete: Function,
+};
+
+// "Let's kick it up a notch!" - Emeril Lagasse
+
+type Props = {
+  todo: Object,
+  onComplete: (todo: Todo) => any,
+};
+{% endhighlight %}
+
+At this point, we've defined the property `onComplete` to be a function that takes a `Todo` as its first and only argument. We don't actually care what that function returns, it's simply a callback, and those guys are super forgettable, so we specify `any` as the return type. With this in place, let's run `flow check` and see what we get:
+
+<div style="text-align:center;margin: 1rem auto;">
+  <img src="http://i.imgur.com/wBsGhTM.png"/>
```

Comment:
Preferably we would store all images inside this app to keep all of the content consolidated in one place.